Client/Frontend/Browser/BrowserViewController.swift

```diff
@@ -847,6 +847,48 @@ extension BrowserViewController: BrowserToolbarDelegate {
     }
 }
 
+protocol WindowCloseHelperDelegate: class {
+    func windowCloseHelper(windowCloseHelper: WindowCloseHelper, didRequestToCloseBrowser browser: Browser)
+}
+
+class WindowCloseHelper: BrowserHelper {
+    weak var delegate: WindowCloseHelperDelegate?
+    private weak var browser: Browser?
+
+    required init(browser: Browser) {
+        self.browser = browser
+        if let path = NSBundle.mainBundle().pathForResource("WindowCloseHelper", ofType: "js") {
+            if let source = NSString(contentsOfFile: path, encoding: NSUTF8StringEncoding, error: nil) as? String {
+                var userScript = WKUserScript(source: source, injectionTime: WKUserScriptInjectionTime.AtDocumentEnd, forMainFrameOnly: true)
+                browser.webView!.configuration.userContentController.addUserScript(userScript)
+            }
+        }
+    }
+
+    func scriptMessageHandlerName() -> String? {
+        return "windowCloseHelper"
+    }
+
+    func userContentController(userContentController: WKUserContentController, didReceiveScriptMessage message: WKScriptMessage) {
+        println("This is WindowCloseHelper.userContentController(_:, didReceiveScriptMessage:)")
+        if let browser = browser {
+            dispatch_async(dispatch_get_main_queue()) {
```

Comment:
But we're always guaranteed to be on the main thread here, right? I think the callbacks for any view are always run on the UI thread.